File: python/erghi/hmac.py
"""
Erghi Python SDK — HMAC utilities (server-side)

Stateless module-level helpers for generating visitor identity hashes and
verifying webhook signatures. They do not require an ``ErghiClient``
instance. Keep your widgetSecretKey / webhook secret on the server —
never expose it to clients.
"""

from __future__ import annotations

import hashlib
import hmac as hmac_lib


def generate_identity_hash(visitor_id: str, secret_key: str) -> str:
    """
    Generate an HMAC-SHA256 hex digest for a visitor ID.

    Call this server-side to securely link your logged-in user to the Erghi widget.

    Args:
        visitor_id: Your system's user/visitor identifier.
        secret_key: The ``whsec_...`` secret from the Erghi Admin Portal.

    Returns:
        Lowercase hex-encoded HMAC-SHA256 digest (64 characters).

    Raises:
        ValueError: If ``visitor_id`` or ``secret_key`` is empty.

    Example::

        from erghi import generate_identity_hash

        identity_hash = generate_identity_hash(
            visitor_id=str(request.user.id),
            secret_key=settings.ERGHI_WIDGET_SECRET,
        )
    """
    if not visitor_id or not secret_key:
        raise ValueError("visitor_id and secret_key are required")

    digest = hmac_lib.new(
        key=secret_key.encode("utf-8"),
        msg=visitor_id.encode("utf-8"),
        digestmod=hashlib.sha256,
    ).hexdigest()
    return digest


def verify_webhook_signature(payload: str, signature: str, secret: str) -> bool:
    """
    Verify an incoming Erghi webhook signature.

    Erghi signs webhook payloads with ``HMAC-SHA256(raw_body, webhook_secret)``
    and sends the result in the ``X-Erghi-Signature`` header. Always verify
    this before processing webhook events.

    Args:
        payload: The raw request body string (do not parse JSON first).
        signature: The value of the ``X-Erghi-Signature`` header.
        secret: Your workspace webhook secret.

    Returns:
        ``True`` if the signature is valid, ``False`` otherwise.
    """
    if not payload or not signature or not secret:
        return False

    try:
        expected = generate_identity_hash(payload, secret)
        # Constant-time comparison to prevent timing attacks
        return hmac_lib.compare_digest(expected.lower(), signature.lower())
    except Exception:
        return False

File: javascript/src/hmac.ts
/**
 * HMAC utilities for server-side visitor identity verification and webhook
 * signature validation.
 *
 * These are stateless module-level helpers — they do not require an
 * `ErghiClient` instance. Only call them from your backend; never ship your
 * widget secret key or webhook secret to the browser.
 */
import { createHmac, timingSafeEqual } from 'crypto';

/**
 * Generates an HMAC-SHA256 hex digest of the visitorId using the widgetSecretKey.
 *
 * Call this from your backend when a user is logged in and you want to pass
 * their identity to the Erghi widget securely.
 *
 * @param visitorId - Your system's user/visitor identifier (e.g. a UUID or user ID)
 * @param secretKey - The `whsec_...` secret key from the Erghi Admin Portal
 * @returns Lowercase hex-encoded HMAC-SHA256 digest (64 characters)
 *
 * @example
 * ```ts
 * // On your server:
 * const hash = generateIdentityHash(currentUser.id, process.env.ERGHI_WIDGET_SECRET!);
 * // Pass `hash` and `currentUser.id` to your frontend for use with the widget
 * ```
 */
export function generateIdentityHash(visitorId: string, secretKey: string): string {
  if (!visitorId || !secretKey) {
    throw new Error('visitorId and secretKey are required');
  }

  return createHmac('sha256', secretKey).update(visitorId, 'utf8').digest('hex');
}

/**
 * Verifies an incoming Erghi webhook signature.
 *
 * Erghi signs webhook payloads with `HMAC-SHA256(rawBody, webhookSecret)` and
 * sends the result in the `X-Erghi-Signature` header. Always verify this
 * before processing webhook events.
 *
 * @param payload - The raw request body string (do not parse JSON first)
 * @param signature - The value of the `X-Erghi-Signature` header
 * @param secret - Your workspace webhook secret
 * @returns `true` if the signature is valid, `false` otherwise
 */
export function verifyWebhookSignature(payload: string, signature: string, secret: string): boolean {
  if (!payload || !signature || !secret) {
    return false;
  }

  try {
    const expected = generateIdentityHash(payload, secret);
    const expectedBuffer = Buffer.from(expected, 'utf8');
    const signatureBuffer = Buffer.from(signature.toLowerCase(), 'utf8');

    if (expectedBuffer.length !== signatureBuffer.length) {
      return false;
    }

    return timingSafeEqual(expectedBuffer, signatureBuffer);
  } catch {
    return false;
  }
}
